### ISSUE TITLE
Faulty numpy randomness when using GPU

### ISSUE THREAD

**Issue by nikonikolov** [tensorflow/tensorflow#18874]
### System information
- Have I written custom code: Yes
- OS Platform and Distribution: Tested on Slackware Linux 14.2 and Ubuntu 16.04
- TensorFlow installed from: binary
- TensorFlow version: tested on both 1.4 and 1.6
- Python version: 3.6
- CUDA/cuDNN version: 8.0 for TF 1.4 and 9.0 for TF 1.6
- Bazel version: N/A
- GPU model and memory: tested on GTX 960M and GTX 1080
- Exact command to reproduce: N/A

### Describe the problem
I am unable to reproduce random numbers generated from numpy when I use it in combination with TF. In the beginning of all my tests, I set
```
tf.set_random_seed(seed)
np.random.seed(seed)
```

I have been debugging, and when I use numpy and no TF, all results are reproducible. When I add the TF code, the random numbers stop being reproducible. When I use both TF and numpy, I get the following results:

1. TF variables are initialized to the same value every time (OK)
2. When I use `np.random.RandomState()` with a set seed instead of direct calls to `np.random.uniform()`, `np.random.normal()`, etc, results are reproducible (OK)
3. When I use direct calls to `np.random.uniform()`, `np.random.normal()`, etc, results are reproducible on CPU but not on GPU (NOT OK)
1080
Since the results are reproducible when using CPU but not GPU, it made me think that this might be a possible bug. 

I am not using any threads so the problem is definitely not caused by race conditions. I am monitoring reproducibility of results only by the random numbers which are generated, which are not in any way affected by the training results from the TF neural net. What is really strange is that the piece of code that seems to be affecting the results is the part about computing and backpropagating gradients. I do not expect that this uses any random numbers generated by numpy in the backend. Furthermore, even if it did, the order of my calls to `np.random` and to `sess.run` is always deterministic, so the same random numbers should be observed between separate runs. 

My code is somewhat too big at the moment to post. I can try to compile some simple example where the issue occurs, but I first wanted to make sure that this is indeed not the expected behavior.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
Bazel version
GPU model and memory
Exact command to reproduce

**Comment by nikonikolov** [comment]
Thanks for the response. I think you misunderstood what I was saying, maybe I was not clear enough.
Here is a mock up (just to illustrate the issue, it is not the *exact* piece of code which reproduces the problem):

```
seed = 42
np.random.seed(seed)
tf.set_random_seed(seed)

# build graph
sess = tf.Session()
rands = []

for i in range(30000):
  r = np.random.uniform(0,1)
  rands.append(r)
  sess.run(train_op, feed_dict=...)
```

The issue that I am experiencing is that `rands` ends up with different numbers between runs. This is happening only when using GPU. On CPU, the numbers are the same. When I remove the line `sess.run(train_op, feed_dict=...)`, everything works fine. Also if I use `prng = np.random.RandomState(); prng.seed(seed)` and then sample with `r = prng.uniform(0,1)`, the generated numbers are the same between runs on both CPU and GPU (`sess.run(train_op, feed_dict=...)` is also being executed).

I was wondering whether this is expected behavior? If not, I will try to create a proper example which reproduces the problem so we can debug further.

**Comment by reedwm** [comment]
So the issue is that running TensorFlow sessions seems to affect the random numbers that numpy generates.

/CC @zffchen78 do you know if this is expected? IMO this shouldn't occur, because TensorFlow should only use it's own seed for randomness.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zffchen78: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zffchen78: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zffchen78: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zffchen78: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zffchen78: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zffchen78: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zffchen78: It has been 104 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zffchen78: It has been 125 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zffchen78: It has been 140 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zffchen78: It has been 155 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zffchen78: It has been 170 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zffchen78: It has been 200 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zffchen78: It has been 215 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jvishnuvardhan** [comment]
@nikonikolov Could check whether the error persists with TF1.12 or 1.13? Thanks!

**Comment by jvishnuvardhan** [comment]
Closing due to lack of recent activity, but please let me know if I'm mistaken. Since this issue is old at this point, please reopen the issue if it still occurs when tried with the latest version of Tensorflow. Thank you.
